docs/proposals/202104_native_support_of_multitenancy.md

```diff
@@ -0,0 +1,83 @@
+* https://github.com/thanos-io/thanos/issues/3822 (Ability to require tenant in queries)
+* https://github.com/thanos-io/thanos/issues/3819 (Thanos receive per tenant limits)
+* https://github.com/thanos-io/thanos/issues/3820 (Zone aware replication)
+* https://github.com/thanos-io/thanos/issues/3572 (Track data usage across tenant / system)
+* https://github.com/thanos-io/thanos/issues/1318 (*: Support multi-tenancy on object storage using custom prefix)
+* https://github.com/thanos-io/thanos/pull/3289 (receiver/compact/store/sidecar: Added a provision in the bucket config to add a prefix)
+
+### Summary
+
+We want to introduce a special tenant label for identifying a tenant and make each component aware of it and then finally track the data usages across the tenants. The receiver is on the safe side as it has a tenant-specific label to uniquely identify tenants.
+
+All the components implementing either the Store API (Sidecar, Ruler, Store Gateway, Receiver) or Query API (Thanos querier frontend and Thanos querier) will be affected. Currently, the ruler is exposing the store API but since there is an ongoing LFX project (to make the ruler stateless), it will upload the data to the receiver itself and it might no longer expose Store API.
+
+### Motivation
+
+* We actually don't have any way to track the resource usages of the query(like error rates, memory allocation, CPU usages along with the queries) and tenant-wise metrics.
+* Every component should allow passing a special label that allows APIs that identify tenants.
+* Currently, we use external labels to achieve multi-tenancy. These labels are sufficient for the sidecar (single-tenant per sidecar i.e hard tenant) but for other components (where we will have multiple tenants per component i.e soft tenancy) these external labels are not sufficient.
+
+### Goals
+
+1. Introduce a mechanism to identify tenants to support multi-tenancy use cases.
+2. Isolate all the incoming query requests.
+3. Track data usage across tenants.
+
+#### Use Case
+
+1. Allow admins to obtain tenant information on per tenant queries, operations, and ingestion.
+2. Sending tenant-specific alerts to different alert managers.
+3. Support cross tenant views.
+
+### Non Goals
+
+*  Not introduce any type of authorization/authenticatication mechanism.
+    * We should deal with auth outside Thanos using some authentication proxy (eg. kube-rbac-proxy) allowing us to use any authN/authZ system we want.
+    * Evaluate some set of rules and check if a certain label can access the data or not.
+
+### Solution
+
+* Introduce a new flag to enable or disable the multi-tenancy mode.
+* When this flag is enabled we will modify the metrics which can be used to track usage characteristics of a tenant by injecting the tenant’s label (a const. Label ).
+* We will make sure that the query request received has a tenant label and if not we have some default tenant label configured.
```

Comment:
Why not blocking if tenant is not specified?

Reply:
we can but we can also have a default tenant label to use when none is specified.

Reply:
If we have switched on multi-tenancy mode, let's have no default label. Users know that if they have enabled multitenancy, they should provide a label for that, and not depend on a default tenant label. If they want a default behaviour, then it does not make sense to have multitenancy enabled. WDYT?

Reply:
Yes, I think that would be good. But we need to ensure how we do that on Receiver too (it would be good to do so)